File: api/v1/endpoints/data_acquisition.py
# -*- coding: utf-8 -*-
"""One-stop model-assisted data acquisition endpoints."""

from io import BytesIO
from datetime import date, timedelta

from fastapi import APIRouter, HTTPException, Query
from fastapi.responses import FileResponse, StreamingResponse

from api.v1.schemas.data_acquisition import (
    DataAcquisitionPlanRequest,
    DataAcquisitionRunRequest,
    ResearchReportExportRequest,
)
from src.services.data_acquisition_service import DataAcquisitionError, DataAcquisitionService
from src.services.data_acquisition_task_service import DataAcquisitionTaskService
from src.services.research_report_library_service import (
    ResearchReportLibraryError,
    ResearchReportLibraryService,
)

router = APIRouter()


def _error(exc: Exception, status_code: int = 400) -> HTTPException:
    return HTTPException(status_code=status_code, detail={"error": "data_acquisition_error", "message": str(exc)[:500]})


@router.get("/capabilities", summary="列出一站式取数可用渠道与能力")
def capabilities():
    return DataAcquisitionService().capabilities()


@router.post("/plan", summary="由大模型把自然语言需求拆成可审计取数计划")
def plan(request: DataAcquisitionPlanRequest):
    try:
        return DataAcquisitionService().plan(request.request)
    except DataAcquisitionError as exc:
        raise _error(exc, 502)


@router.post("/run", summary="执行确认后的多渠道取数计划并生成统一数据包")
def run(request: DataAcquisitionRunRequest):
    try:
        return DataAcquisitionService().run(request.request, request.plan)
    except DataAcquisitionError as exc:
        raise _error(exc)


@router.post("/run-async", status_code=202, summary="提交后台取数任务并返回真实进度编号")
def run_async(request: DataAcquisitionRunRequest):
    try:
        return DataAcquisitionTaskService.get_instance().submit(request.request, request.plan)
    except DataAcquisitionError as exc:
        raise _error(exc)


@router.get("/tasks/{task_id}", summary="读取后台取数任务的真实渠道与打包进度")
def task(task_id: str):
    try:
        return DataAcquisitionTaskService.get_instance().get(task_id)
    except DataAcquisitionError as exc:
        raise _error(exc, 404)


@router.get("/jobs", summary="列出最近生成的数据包")
def jobs(limit: int = Query(20, ge=1, le=100)):
    return DataAcquisitionService().list_jobs(limit)


@router.get("/jobs/{job_id}", summary="读取数据包来源清单与执行结果")
def job(job_id: str):
    try:
        return DataAcquisitionService().get_job(job_id)
    except DataAcquisitionError as exc:
        raise _error(exc, 404)


@router.get("/jobs/{job_id}/download", summary="下载包含 JSON、CSV、Excel 和来源清单的 ZIP")
def download(job_id: str):
    try:
        path = DataAcquisitionService().package_path(job_id)
    except DataAcquisitionError as exc:
        raise _error(exc, 404)
    return FileResponse(path, media_type="application/zip", filename=f"财经数据包_{job_id}.zip")


@router.get("/research-reports/status", summary="读取本地研报库覆盖范围与后台同步进度")
def research_report_status():
    return ResearchReportLibraryService.get_instance().ensure_background_sync()


@router.post("/research-reports/sync", status_code=202, summary="后台补齐最近两年研报元数据与PDF链接")
def sync_research_reports(years: int = Query(2, ge=1, le=5), force: bool = Query(False)):
    service = ResearchReportLibraryService.get_instance()
    end = date.today()
    try:
        return service.start_sync(
            start_date=end - timedelta(days=365 * years),
            end_date=end,
            force=force,
        )
    except ResearchReportLibraryError as exc:
        raise _error(exc)


@router.get("/research-reports/facets", summary="读取研报本地库可人工筛选标签")
def research_report_facets():
    return ResearchReportLibraryService.get_instance().facets()


@router.get("/research-reports/search", summary="只从本地SQLite研报库执行人工条件检索")
def search_research_reports(
    title_query: str = Query("", max_length=200),
    content_query: str = Query("", max_length=500),
    broker: str = Query("", max_length=160),
    company: str = Query("", max_length=160),
    ts_code: str = Query("", max_length=20),
    report_type: str = Query("", max_length=80),
    industry: str = Query("", max_length=160),
    author: str = Query("", max_length=160),
    tag: str = Query("", max_length=80),
    start_date: str = Query("", max_length=10),
    end_date: str = Query("", max_length=10),
    has_pdf: bool = Query(True),
    sort: str = Query("latest", pattern="^(latest|oldest)$"),
    page: int = Query(1, ge=1),
    page_size: int = Query(30, ge=1, le=100),
):
    return ResearchReportLibraryService.get_instance().search(
        title_query=title_query, content_query=content_query, broker=broker,
        company=company, ts_code=ts_code, report_type=report_type,
        industry=industry, author=author, tag=tag, start_date=start_date,
        end_date=end_date, has_pdf=has_pdf, sort=sort, page=page, page_size=page_size,
    )


@router.post("/research-reports/export-selected", summary="导出人工勾选研报及PDF链接Excel")
def export_selected_research_reports(request: ResearchReportExportRequest):
    try:
        content = ResearchReportLibraryService.get_instance().export_selected(request.ids)
    except ResearchReportLibraryError as exc:
        raise _error(exc)
    return StreamingResponse(
        BytesIO(content),
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": 'attachment; filename="selected_research_reports.xlsx"'},
    )

File: apps/dsa-web/src/api/dataAcquisition.ts
import apiClient from './index';
import { toCamelCase } from './utils';
import { cachedQuery, invalidateCachedQueries } from './requestCache';
import type {
  AcquisitionCapabilities,
  AcquisitionDownloadProgress,
  AcquisitionJob,
  AcquisitionPlan,
  AcquisitionRunTask,
  ResearchReportFacets,
  ResearchReportLibraryStatus,
  ResearchReportSearchFilters,
  ResearchReportSearchResult,
} from '../types/dataAcquisition';

type RawPlan = Omit<AcquisitionPlan, 'outputFormats' | 'generatedAt' | 'scope'> & {
  output_formats?: string[];
  generated_at?: string;
  include_files?: boolean;
  scope?: { symbols?: string[]; company_names?: string[]; keywords?: string[]; start_date?: string; end_date?: string; market_wide?: boolean };
};

function readPlan(raw: RawPlan): AcquisitionPlan {
  return {
    ...raw,
    tasks: (raw.tasks ?? []).map((task) => ({ ...task, params: task.params ?? {} })),
    outputFormats: raw.output_formats ?? [],
    generatedAt: raw.generated_at ?? '',
    includeFiles: raw.include_files ?? false,
    scope: {
      symbols: raw.scope?.symbols ?? [], companyNames: raw.scope?.company_names ?? [],
      keywords: raw.scope?.keywords ?? [], startDate: raw.scope?.start_date ?? '',
      endDate: raw.scope?.end_date ?? '', marketWide: raw.scope?.market_wide ?? false,
    },
  };
}

function writePlan(plan: AcquisitionPlan) {
  return {
    title: plan.title, objective: plan.objective, model: plan.model, caveats: plan.caveats,
    generated_at: plan.generatedAt, output_formats: plan.outputFormats, include_files: plan.includeFiles ?? false,
    scope: { symbols: plan.scope.symbols, company_names: plan.scope.companyNames, keywords: plan.scope.keywords,
      start_date: plan.scope.startDate, end_date: plan.scope.endDate, market_wide: plan.scope.marketWide },
    // params is an opaque upstream contract and must never be camel-cased.
    tasks: plan.tasks.map((task) => ({ ...task, params: task.params })),
  };
}

export const dataAcquisitionApi = {
  capabilities: async (): Promise<AcquisitionCapabilities> => {
    return cachedQuery('acquisition:capabilities', async () => {
      const response = await apiClient.get('/api/v1/data-acquisition/capabilities');
      return toCamelCase<AcquisitionCapabilities>(response.data);
    }, { freshMs: 5 * 60_000, staleMs: 24 * 60 * 60_000 });
  },
  plan: async (request: string): Promise<AcquisitionPlan> => {
    const response = await apiClient.post('/api/v1/data-acquisition/plan', { request }, { timeout: 150000 });
    return readPlan(response.data as RawPlan);
  },
  runAsync: async (request: string, plan: AcquisitionPlan): Promise<AcquisitionRunTask> => {
    const response = await apiClient.post('/api/v1/data-acquisition/run-async', { request, plan: writePlan(plan) });
    return toCamelCase<AcquisitionRunTask>(response.data);
  },
  task: async (taskId: string): Promise<AcquisitionRunTask> => {
    const response = await apiClient.get(`/api/v1/data-acquisition/tasks/${encodeURIComponent(taskId)}`);
    const task = toCamelCase<AcquisitionRunTask>(response.data);
    if (task.result) task.result.plan = readPlan((response.data as { result: { plan: RawPlan } }).result.plan);
    return task;
  },
  jobs: async (): Promise<{ items: AcquisitionJob[]; total: number }> => {
    return cachedQuery('acquisition:jobs', async () => {
      const response = await apiClient.get('/api/v1/data-acquisition/jobs');
      return toCamelCase(response.data);
    }, { freshMs: 3_000, staleMs: 15_000 });
  },
  download: async (
    jobId: string,
    onProgress?: (progress: AcquisitionDownloadProgress) => void,
  ): Promise<Blob> => {
    const response = await apiClient.get(`/api/v1/data-acquisition/jobs/${encodeURIComponent(jobId)}/download`, {
      responseType: 'blob', timeout: 120000,
      onDownloadProgress: (event) => {
        const total = event.total && event.total > 0 ? event.total : undefined;
        onProgress?.({
          loaded: event.loaded,
          total,
          percent: total ? Math.min(100, Math.round((event.loaded / total) * 100)) : undefined,
        });
      },
    });
    return response.data as Blob;
  },
  researchReportStatus: async (): Promise<ResearchReportLibraryStatus> => {
    return cachedQuery('acquisition:research-status', async () => {
      const response = await apiClient.get('/api/v1/data-acquisition/research-reports/status');
      return toCamelCase(response.data);
    }, { freshMs: 5_000, staleMs: 30_000 });
  },
  syncResearchReports: async (years = 2): Promise<ResearchReportLibraryStatus> => {
    const response = await apiClient.post('/api/v1/data-acquisition/research-reports/sync', undefined, { params: { years } });
    invalidateCachedQueries('acquisition:research-');
    return toCamelCase(response.data);
  },
  researchReportFacets: async (): Promise<ResearchReportFacets> => {
    return cachedQuery('acquisition:research-facets', async () => {
      const response = await apiClient.get('/api/v1/data-acquisition/research-reports/facets');
      return toCamelCase(response.data);
    }, { freshMs: 10 * 60_000, staleMs: 24 * 60 * 60_000 });
  },
  searchResearchReports: async (
    filters: ResearchReportSearchFilters,
    page = 1,
    pageSize = 30,
  ): Promise<ResearchReportSearchResult> => {
    const key = `acquisition:research-search:${JSON.stringify({ filters, page, pageSize })}`;
    return cachedQuery(key, async () => {
      const response = await apiClient.get('/api/v1/data-acquisition/research-reports/search', {
        params: {
          title_query: filters.titleQuery, content_query: filters.contentQuery,
          broker: filters.broker, company: filters.company, ts_code: filters.tsCode,
          report_type: filters.reportType, industry: filters.industry, author: filters.author,
          tag: filters.tag, start_date: filters.startDate, end_date: filters.endDate,
          has_pdf: filters.hasPdf, sort: filters.sort, page, page_size: pageSize,
        },
      });
      return toCamelCase(response.data);
    }, { freshMs: 30_000, staleMs: 5 * 60_000 });
  },
  exportSelectedResearchReports: async (ids: number[]): Promise<Blob> => {
    const response = await apiClient.post('/api/v1/data-acquisition/research-reports/export-selected', { ids }, {
      responseType: 'blob', timeout: 120000,
    });
    return response.data as Blob;
  },
};
